File: algorithms/318 - Maximum Product of Word Lengths.js
/**
 * @param {string[]} words
 * @return {number}
 */
var maxProduct = function(words) {
    let ret = 0;

    for (let i = 0; i < words.length; i++) {
        for (let j = i + 1; j < words.length; j++) {
            if (checkInteresction(words[i], words[j])) ret = Math.max(ret, words[i].length * words[j].length);
        }
    }

    return ret;
};

const checkInteresction = (word1, word2) => {
    for (let i = 0; i < word1.length; i++) {
        if (word2.includes(word1[i])) return false;
    }

    return true;
}

File: algorithms/ruby/318 - Maximum Product of Word Lengths.rb
# @param {String[]} words
# @return {Integer}
def max_product(words)
  words.sort_by!(&:length).reverse!
  max = 0
  for i in 0..(words.length - 1) do
    for j in (i + 1)..(words.length - 1) do
      if words[i].count(words[j]) == 0
        max = [max, words[i].length * words[j].length].max
      end
    end
  end
  max
end
